Crashlytics/Crashlytics/Models/Record/FIRCLSRecordAdapter.m

```diff
@@ -0,0 +1,147 @@
+  self.binaryImages = [FIRCLSRecordBinaryImage
+      binaryImagesFromDictionaries:[FIRCLSRecordAdapter dictionariesFromEachLineOfFile:path]];
+}
+
+- (void)loadMetaDataFile {
+  NSString *path = [self.folderPath stringByAppendingPathComponent:CLSReportMetadataFile];
+  NSDictionary *dict = [FIRCLSRecordAdapter combinedDictionariesFromFilePath:path];
+
+  self.identity = [[FIRCLSRecordIdentity alloc] initWithDict:dict[@"identity"]];
+  self.host = [[FIRCLSRecordHost alloc] initWithDict:dict[@"host"]];
+  self.application = [[FIRCLSRecordApplication alloc] initWithDict:dict[@"application"]];
+  self.executable = [[FIRCLSRecordExecutable alloc] initWithDict:dict[@"executable"]];
+}
+
+- (void)loadSignalFile {
+  NSString *path = [self.folderPath stringByAppendingPathComponent:CLSReportSignalFile];
+  NSDictionary *dicts = [FIRCLSRecordAdapter combinedDictionariesFromFilePath:path];
+
+  self.signal = [[FIRCLSRecordSignal alloc] initWithDict:dicts[@"signal"]];
+  self.runtime = [[FIRCLSRecordRuntime alloc] initWithDict:dicts[@"runtime"]];
+  self.processStats = [[FIRCLSRecordProcessStats alloc] initWithDict:dicts[@"process_stats"]];
+  self.storage = [[FIRCLSRecordStorage alloc] initWithDict:dicts[@"storage"]];
+  self.threads = [FIRCLSRecordThread threadsFromDictionaries:dicts[@"threads"]
+                                                   withNames:dicts[@"thread_names"]
+                                                 withRuntime:self.runtime];
+}
+
+- (void)loadKeyValuesFile {
+  NSString *path =
+      [self.folderPath stringByAppendingPathComponent:CLSReportInternalIncrementalKVFile];
+  self.keyValues = [FIRCLSRecordKeyValue
+      keyValuesFromDictionaries:[FIRCLSRecordAdapter dictionariesFromEachLineOfFile:path]];
+}
+
+/// Return the persisted crash file as a combined dictionary that way lookups can occur with a key
+/// (to avoid ordering dependency)
+/// @param filePath Persisted crash file path
++ (NSDictionary *)combinedDictionariesFromFilePath:(NSString *)filePath {
+  NSMutableDictionary *joinedDict = [[NSMutableDictionary alloc] init];
+  for (NSDictionary *dict in [self dictionariesFromEachLineOfFile:filePath]) {
+    [joinedDict addEntriesFromDictionary:dict];
+  }
+  return joinedDict;
+}
+
+/// The persisted crash files contains JSON on separate lines. Read each line and return the JSON
+/// data as a dictionary.
+/// @param filePath Persisted crash file path
++ (NSArray<NSDictionary *> *)dictionariesFromEachLineOfFile:(NSString *)filePath {
+  NSString *content = [[NSString alloc] initWithContentsOfFile:filePath
+                                                      encoding:NSUTF8StringEncoding
+                                                         error:nil];
+  NSArray *lines =
+      [content componentsSeparatedByCharactersInSet:NSCharacterSet.newlineCharacterSet];
+
+  NSMutableArray<NSDictionary *> *array = [[NSMutableArray<NSDictionary *> alloc] init];
+
+  int lineNum = 1;
```

Comment:
Should this start at 0?

Reply:
I am gonna keep this starting at 1. When using text editors, they usually don't have the lines starting at 0.